### DISCUSSION TITLE
Homebrew on linux without root access

### DISCUSSION THREAD

**Discussion by DE0CH** [Homebrew/discussions#3386] [Linux] [answered]
My university has some linux machines in the computer lab, but a lot of packages are outdated. I, a student, can't update them because I don't have root access and talking to admin is not preferred because that involves human to human interaction. So I've been using homebrew to build and install packages from source. It's working like a charm so far. Just wondering if this is the orthodox way to install packages on machines without root access.

**Comment by carlocab** [comment]
It might be easier to ask your admin to create a `linuxbrew` user and install Homebrew into the `linuxbrew` user's home directory. This would require talking to an admin, but you should only need to do this once. In return, you would be able to avoid having to build packages from source.

@Homebrew/linux

**Comment by dawidd6** [comment, marked as answer]
You can make use of `HOMEBREW_RELOCATE_BUILD_PREFIX` environment variable. Just set it in your shell profile and then install Homebrew in your HOME using instructions from https://docs.brew.sh/Homebrew-on-Linux.

Note that the feature behind this environment variable is still in "testing" phase, so breakages could happen. Yet I've been using it since a couple of months and it works fine, no building from source happened for me just yet.

**Reply by danielnachun** [reply]
>It might be easier to ask your admin to create a `linuxbrew` user and install Homebrew into the `linuxbrew` user's home directory. This would require talking to an admin, but you should only need to do this once. In return, you would be able to avoid having to build packages from source.

In my experience if the machines are multi-user, this wouldn't be allowed because there's be no way for individual users to manage their packages separately.

> You can make use of `HOMEBREW_RELOCATE_BUILD_PREFIX` environment variable. Just set it in your shell profile and then install Homebrew in your HOME using instructions from https://docs.brew.sh/Homebrew-on-Linux.
> 
> Note that the feature behind this environment variable is still in "testing" phase, so breakages could happen. Yet I've been using it since a couple of months and it works fine, no building from source happened for me just yet.

The only caveat to `HOMEBREW_RELOCATE_BUILD_PREFIX` is that currently, the length of your prefix for Homebrew must be less than or equal to the length of `/home/linuxbrew/.linuxbrew`.  Since originally implementing it, I've found a few packages don't quite work with it - `glibc`, `fontconfig` and possibly `go`, `ruby` and `perl`.  Eventually my goal is to rebuild all packages that are currently not relocatable with a very long prefix (255 characters) so that relocation works in all prefixes, and to fix packages like `glibc` and `fontconfig` that don't automatically work with the relocation, though there is no timeline as to when these things will be fixed.

**Reply by DE0CH** [reply]
> You can make use of `HOMEBREW_RELOCATE_BUILD_PREFIX` environment variable. Just set it in your shell profile and then install Homebrew in your HOME using instructions from https://docs.brew.sh/Homebrew-on-Linux.

Thanks. I didn't know this. I just cloned the homebrew repo to a folder in my home directory and exported the `brew shellenv` variables. Would setting this allow me to install formulas without building them from source? 

Also the gcc on the computers are a bit broken, so I installed gcc with `--force-bottle`. It works, but is there a better way to do it?

**Reply by danielnachun** [reply]
> Thanks. I didn't know this. I just cloned the homebrew repo to a folder in my home directory and exported the `brew shellenv` variables. Would setting this allow me to install formulas without building them from source?

Assuming your prefix is short enough, yes (it will fail with an error if your prefix is too long).  As mentioned above, some formula will break when you do this, but most should just work.

> Also the gcc on the computers are a bit broken, so I installed gcc with `--force-bottle`. It works, but is there a better way to do it?

`gcc` is supposed to be relocatable, so this is totally fine.  I need to update the fix I added to `brew` to make GCC relocatable, and then it will also be installed from a bottle regardless of prefix.

**Reply by xiaoran007** [reply]
Works, thank you!

**Comment by tddschn** [comment]
Hi, I try to run 
`HOMEBREW_RELOCATE_BUILD_PREFIX=~/l /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"` 
but the installation script still attempts to install in `/home/linuxbrew/.linuxbrew` and fails because of having no root access, not `~/l` (which, when `~` is expanded, the full path is equal to the length of `/home/linuxbrew/.linuxbrew`.

Could someone help me with this issue? Thanks.

**Reply by DE0CH** [reply]
I think that environment variable is gone (I can't find it anywhere). According to the [ducomentation](https://docs.brew.sh/Installation#untar-anywhere-unsupported), you can do the following

```bash
mkdir ~/l && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C ~/l
eval "$(~/l/bin/brew shellenv)"
brew update --force --quiet
chmod -R go-w "$(brew --prefix)/share/zsh"
```
